### DISCUSSION TITLE
ЛР 6 для ИВТ. Проблемы с регистром mie и инструкцией mret в RARS

### DISCUSSION THREAD

**Discussion by SignComet** [MPSU/exAPS#28] [Вопросы и ответы] [answered]
При использовании RARS - инструмента, который умеет работать с инструкциями для CSR, обнаружила проблему, что он воспринимает сами инструкции, но не воспринимает регистры mie, mtvec и тд. Если в данной программе справа открыть список регистров(Control and Status), то можно увидеть utvec, uim и тд(User).  Заходила в HELP->exception там тоже не нашла решение(тоже примеры в user режиме).  В лабораторной было сказано про переключение между режимами с помощью специальной инструкции ecall. 

Код:       li      t0,    0x00000028
              csrw  t0,   mie              # загружаем маску в регистр маски

Error:    "mie": operand is of incorrect type

**Comment by gr33nka** [comment, marked as answer]
`ecall` нужен для вызовы исключения (exception - синхронного события), которое переведет процессор в режим работы операционной системы (у нас такое не поддерживается). Про `ecall` в качестве справки дано.

Что касается ответа на вопрос, то попробуй использовать не названия регистров, а их адреса. 

``` assembly risc-v
csrw 0x304, t0 
```

Вроде бы сначала идет адрес CSR, а потом адрес регистра в регистровом файле.

Напиши, пожалуйста, получалось ли

**Reply by gr33nka** [reply]
Забыл, что RARS на Java. Поставил и проверил, да, это рабочий вариант, который генерирует правильные инструкции.

При этом почему-то в RARS порядок написания аргументов действительно отличается, твой вариант оказался правильным:

``` assembly risc-v
  csrw t0, 0x304
```

Когда убедишься отметь, пожалуйста, что это "решение", нажав на соответствующую кнопку

**Reply by SignComet** [reply]
Да, спасибо, так действительно работает, но визуально у нас нет возможности увидеть, как происходит запись именно в этот регистр(mie), например, как мы можем видеть ячейки регистров общего назначения, верно?

**Reply by gr33nka** [reply]
Не нашел как это сделать, как будто бы нельзя(

Могу предложить костыль: написать программу для `user` регистров, отладить, а потом заменить на правильные адреса

**Comment by SignComet** [comment]
Также сохраняются проблемы с использованием инструкции mret, которую ассемблер не воспринимает как инструкцию 
Error: "mret" is not a recognized operator

**Reply by gr33nka** [reply]
Инструкции `uret` и `mret` отличаются полем `imm`. В нашей реализации мы этого даже не проверяем, смотрим только на поле `funct3`, которое равно `000` и в `mret`, и в `uret`. Не проверяем `imm` мы сознательно – у нас нет других режимов, кроме машинного, и операционки нет, поэтому допущение хоть и отходит от спецификации, но на работе никак не скажется.

Строго говоря это разные инструкции, но в нашей реализации устройство управления схавает их как одинаковые. Используй `uret` вместо `mret` и все должно работать.

Когда дойду до этой лабы постараюсь что-нибудь придумать менее костыльное, но пока так, сорян

**Reply by SignComet** [reply]
Спасибо огромное)

**Comment by SignComet** [comment]
Простите, заметила, что команда assemble в RARS действительно выполняется, но при "компиляции" всё равно возникает ошибка и дальше первой инструкции, которая работает с машинными csr не компилирует дальше.
Код:     li        t0, 0x00000028     
            csrw   t0, 0x304
            li        t0, 0xEFFFFFFC 
            csrw   t0, 0x340  

В окне messages:
Error: Attempt to access unavailable CSR
Go: execution terminated with errors.

**Reply by gr33nka** [reply]
Эх, видимо он ругается потому что у него нет окошечка в котором он это мог бы показать это (или более строго - его модель не предполагает их использование). То есть как бы говорит:

> Компилировать хочешь? Да пожалуйста, вот твой машинный код! Но отлаживать я тебе не позволю, потому что я не создаю (не моделирую) в своей памяти регистр по адресу 0x304. Не моделирую, значит его нет, значит запустить твой код не могу

Видимо придется воспользоваться костылем о котором выше писал, сначала для `uie` и остальных `user` регистров замоделить, а перед продакшеном заменить все эти регистры на нужные адреса 🤷‍♂️🤷‍♂️

**Reply by SignComet** [reply]
Хорошо, ещё раз спасибо)
